test/test_typing.py

```diff
@@ -0,0 +1,277 @@
+# Copyright 2020-present MongoDB, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Test that each file in mypy_fails/ actually fails mypy, and test some
+sample client code that uses Motor typings.
+"""
+import os
+import sys
+import unittest
+from test.asyncio_tests import AsyncIOTestCase, asyncio_test
+from typing import TYPE_CHECKING, Any, AsyncIterable, Dict, List, Union
+
+from bson import CodecOptions
+from bson.raw_bson import RawBSONDocument
+from bson.son import SON
+from pymongo.operations import DeleteOne, InsertOne, ReplaceOne
+from pymongo.read_preferences import ReadPreference
+
+from motor.core import AgnosticClient, AgnosticCollection
+
+try:
+    from bson import ObjectId
+    from typing_extensions import NotRequired, TypedDict
+
+    class Movie(TypedDict):
+        name: str
+        year: int
+
+    class MovieWithId(TypedDict):
+        _id: ObjectId
+        name: str
+        year: int
+
+    class ImplicitMovie(TypedDict):
+        _id: NotRequired[ObjectId]
+        name: str
+        year: int
+
+except ImportError:
+    Movie = dict  # type:ignore[misc,assignment]
+    ImplicitMovie = dict  # type: ignore[assignment,misc]
+    MovieWithId = dict  # type: ignore[assignment,misc]
+    TypedDict = None
+    NotRequired = None  # type: ignore[assignment]
+
+
+try:
+    from mypy import api
+except ImportError:
+    api = None
+
+sys.path[0:0] = [""]
+
+
+TEST_PATH = os.path.join(os.path.dirname(os.path.realpath(__file__)), "mypy_fails")
```

Comment:
Could you add the equivalent TestMypyFails tests from pymongo too? https://github.com/mongodb/mongo-python-driver/blob/0d44783eddec96a9ae1a03b148cb4f049dca3e49/test/test_typing.py#L87

Reply:
The tests fail due to the `client.test.test` pattern being of type Any rather than `Collection`. Is this a possible complication of our metaprogramming?

For example:

```
    client: AgnosticClient = AgnosticClient()
    coll: AgnosticCollection = client.test.test
    await coll.insert_many(
        {"a": 1}
    )  # error: Dict entry 0 has incompatible type "str": "int"; expected "Mapping[str, Any]": "int"
```
fails to typecheck as expected

But:

```
    client: AgnosticClient = AgnosticClient()
    await client.test.test.insert_many(
        {"a": 1}
    )  # error: Dict entry 0 has incompatible type "str": "int"; expected "Mapping[str, Any]": "int"
```
successfully typechecks.

Reply:
The problem is that the Client/DB/Collection classes need to return the appropriate type from `__getattr__`/`__getitem__`, not Any.